File: rust-2022/src/day16.rs
use std::collections::HashMap;

type GraphNode = (u8, Vec<u8>);
const N: usize = 51;
const P: usize = 15;
type Paths = [[(u8, u8); P]; N];

fn allpaths(graph: &[GraphNode]) -> Paths {
    let mut vv = [[99u8; N]; N];
    for (i, (_, conn)) in graph.iter().enumerate() {
        for &c in conn.iter() {
            vv[i][c as usize] = 1;
        }
    }
    for k in 0..N {
        for i in 0..N {
            for j in 0..N {
                vv[i][j] = vv[i][j].min(vv[i][k] + vv[k][j]);
            }
        }
    }
    let mut ret = [[(0u8, 0u8); P]; N];
    for (i, v) in vv.iter().enumerate() {
        for (j, r) in v
            .iter()
            .enumerate()
            .filter_map(|(node, time)| {
                (graph[node as usize].0 > 0).then_some((node as u8, *time + 1))
            })
            .enumerate()
        {
            ret[i][j] = r;
        }
    }
    ret
}

pub fn read(inp: &[u8]) -> (Vec<GraphNode>, Paths) {
    let mapping: HashMap<&[u8], u8> = inp
        .split(|&c| c == b'\n')
        .enumerate()
        .map(|(i, line)| (&line[6..8], i as u8))
        .collect();
    let mut graph = Vec::with_capacity(64);
    for line in inp.split(|&c| c == b'\n') {
        let rate = if line[24] == b';' {
            line[23] - b'0'
        } else {
            10 * (line[23] - b'0') + (line[24] - b'0')
        };
        let offset = if line[47] == b' ' { 48 } else { 49 };
        let edges = line[offset..]
            .split(|&c| c == b',')
            .map(|entry| *mapping.get(&entry[(entry[0] == b' ') as usize..]).unwrap() as u8)
            .collect::<Vec<_>>();
        graph.push((rate, edges))
    }
    let paths = allpaths(&graph);
    (graph, paths)
}

fn search(
    node: u8,
    best: u64,
    current: u64,
    time_left: u8,
    enabled: u64,
    paths: &Paths,
    rates: &[u64],
) -> (u64, u64) {
    let mut score = current;
    let mut best = best;
    for &(dst, cur_time) in paths[node as usize].iter() {
        if cur_time > time_left || ((1 << dst) & enabled) != 0 {
            continue;
        }
        let new_time = (time_left - cur_time) as u64;
        let new_score = current + new_time * rates[dst as usize];
        if new_score + new_time * 47 < best {
            continue;
        }
        let x = search(
            dst,
            best,
            new_score,
            new_time as u8,
            enabled | (1 << dst),
            paths,
            rates,
        );
        best = x.1;
        score = score.max(x.0);
    }
    (score, best.max(score))
}

pub fn part1(inp: &(Vec<GraphNode>, Paths)) -> u64 {
    let rates: Vec<_> = inp.0.iter().map(|x| x.0 as u64).collect();
    search(16, 0, 0, 30, 0, &inp.1, &rates).0
}

pub fn part2(inp: &(Vec<GraphNode>, Paths)) -> u64 {
    let rates: Vec<_> = inp.0.iter().map(|x| x.0 as u64).collect();
    search(16, 0, 0, 30, 0, &inp.1, &rates).0
}

pub fn run() -> (String, String) {
    let inp = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day16.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}

File: 2022/day16.py
import time
from collections import deque
from functools import cache

start = time.time()
with open("../inputs/2022/day16.input") as f:
    inp = {}
    for line in f.read().split("\n"):
        src = line[6:8]
        a, b = line.split("; ")
        rate = int(a[23:])
        edges = b[22:].split(",")
        inp[src] = (rate, list(s.strip() for s in edges))


# This is just lucky, the order we visit things changes the relevant magic numbers
@cache
def paths(node: str) -> dict[str, int]:
    q: deque[tuple[int, str]] = deque([(0, node)])
    ret = {}
    while q:
        d, n = q.popleft()
        for dest in inp[n][1]:
            if dest not in ret:
                ret[dest] = d + 2
                q.append((d + 1, dest))
    return {n: t for n, t in ret.items() if inp[n][0] > 0}


def part1(inp: dict) -> int:
    def search(node: str, best: int, current: int, time_left: int, enabled: set[str]):
        score = current
        for dst, cur_time in paths(node).items():
            if dst in enabled or cur_time > time_left:
                continue
            new_time = time_left - cur_time
            new_score = current + new_time * inp[dst][0]
            # magic number for A-* pruning
            if new_score + new_time * 47 < best:
                continue
            candidate, best = search(dst, best, new_score, new_time, enabled | {dst})
            score = max(score, candidate)
        return score, max(best, score)

    return search("AA", 0, 0, 30, set())[0]


def part2(inp: dict) -> int:
    # Could merge with part1 search, but meh
    def search(
        me: str,
        you: str,
        best: int,
        current: int,
        time_left: int,
        enabled: set[str],
        budget: int,
    ):
        score = current
        for dst, cur_time in paths(me).items():
            if dst in enabled or cur_time > time_left:
                continue
            if budget < cur_time:
                new_me, new_you = you, dst
                new_time_left, new_budget = time_left - budget, cur_time - budget
            else:
                new_me, new_you = dst, you
                new_time_left, new_budget = time_left - cur_time, budget - cur_time
            new = current + (time_left - cur_time) * inp[dst][0]
            # magic number for A-* pruning (order dependent for path visiting)
            # so not a general solution
            if new + (time_left - cur_time - budget) * 78 < best:
                continue
            candidate, best = search(
                new_me,
                new_you,
                best,
                new,
                new_time_left,
                enabled | {dst},
                new_budget,
            )
            score = max(score, candidate)
        return score, max(best, score)

    return search("AA", "AA", 0, 0, 26, set(), 0)[0]


print(
    f"Day 16     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)
